### DISCUSSION TITLE
How to handle leak ?

### DISCUSSION THREAD

**Discussion by moulin-louis** [codam-coding-college/MLX42#78] [Q&A]
Hey,
I'm trying to handle leak in my project using MLX42.
I'm destroying all image using mlx_delete_image and I'm calling mlx_terminate at the end of my program on the mlx pointer.
I still have a lot of leak comming from mlx_init(), what should I do and what I am missing ?
Thanks for your help !

**Comment by W2Wizard** [comment]
Hello! Theres not much I can do but if possible could you:
- If tell me which version of MLX42 ? (Pulled recently from master or used a release)
- Provide me some sample code so I can check it for myself.
- Which OS you are using and which version.

If I have a bit more info I can help you track it down.

**Reply by moulin-louis** [reply]
Hey ! Im building MLX42 from his source using the latest commit on ubuntu 22.04.01.

```c
typedef struct s_data
{
	mlx_t		*mlx;
	mlx_image_t	*img;

	mlx_image_t	*nord;
	mlx_image_t	*south;
	mlx_image_t	*west;
	mlx_image_t	*east;
}
```

This is the struct that store my pointer

```c
static	void	clean_exit(t_data *data, int err)
{
	if (data)
	{
		if (data->nord)
			mlx_delete_image(data->mlx, data->nord);
		if (data->south)
			mlx_delete_image(data->mlx, data->south);
		if (data->west)
			mlx_delete_image(data->mlx, data->west);
		if (data->east)
			mlx_delete_image(data->mlx, data->east);
		if (data->img)
			mlx_delete_image(data->mlx, data->img);	
		if (data->mlx)
			mlx_terminate(data->mlx);
		if (data->tmp_map)
			free_array((void **)data->tmp_map);
		if (data->map)
			free_array((void **)data->map);
	}
	exit(err);
}
```

This is the function that is called when I close the windows using the button on the top right.
I just called mlx_init once of course,
if needed all my code is avaible on my repo here :  https://github.com/moulin-louis/cub3d

Thx a lot for your time !!

**Reply by W2Wizard** [reply]
For reference here is what i'm getting:

```bash
Process:         cub3d [16199]
Path:            /private/tmp/*/cub3d
Load Address:    0x1076f6000
Identifier:      cub3d
Version:         0
Code Type:       X86-64
Platform:        macOS
Parent Process:  zsh [15208]

Date/Time:       2023-02-06 15:08:29.925 +0100
Launch Time:     2023-02-06 15:08:02.526 +0100
OS Version:      macOS 12.5.1 (21G83)
Report Version:  7
Analysis Tool:   /Applications/Xcode.app/Contents/Developer/usr/bin/leaks
Analysis Tool Version:  Xcode 14.1 (14B47b)
----

leaks Report Version: 4.0
Process 16199: 22668 nodes malloced for 14088 KB
Process 16199: 2 leaks for 144 total leaked bytes.

    2 (144 bytes) << TOTAL >>
      1 (96 bytes) ROOT LEAK: 0x600000a20000 [96]
      1 (48 bytes) ROOT LEAK: <CFString 0x600002027180> [48]  length: 26  "Copyright Apple Inc., 2019"
```

**Reply by W2Wizard** [reply]
Ok so it seems that causes GLFW the leak which in turn is caused by something thats outside of its control 🙃

**Reply by W2Wizard** [reply]
Lets hope I get something out of it lol

Created an issue:
https://github.com/glfw/glfw/issues/2269

**Reply by moulin-louis** [reply]
Haha glfw is the guilty one here.
Again thank you so much for your time here ! Lets hope the issue will be fix soon

**Reply by W2Wizard** [reply]
Eitherway whenever you see leaks with CFString in there, its alright and not your fault.
